include/random_walks/shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,159 @@
+// VolEsti (volume computation and sampling library)
+
+// Copyright (c) I don't know what to write here :)
+
+// Licensed under GNU LGPL-3.0; see LICENCE file
+
+#ifndef RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+#define RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+
+#include <Eigen/Eigen>
+#include <limits>
+#include <cmath>
+#include <stdexcept>
+
+#include "sampling/sphere.hpp"
+#include "preprocess/feasible_point.hpp"
+#include "convex_bodies/hpolytope.h"
+
+struct ShakeAndBakeWalk
+{
+    enum Mode { Original, Limping, Running };
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point = typename Polytope::PointType;
+        using NT = typename Point::FT;
+        using VT = typename Polytope::VT;        
+
+        Mode mode_{Mode::Original};
+
+        static void set_epsilon(NT new_eps) noexcept { epsilon_ = new_eps; }
+        static NT   get_epsilon()      noexcept      { return epsilon_;   }
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope&       P,
+             RandomNumberGenerator& rng,
+             Mode                   m = Mode::Original)
+            : mode_{m}
+        {
+            initialize(P, rng);
+        }
+
+        template <typename GenericPolytope>
+        inline void apply(GenericPolytope const& P,
+                          unsigned int          walk_len,
+                          RandomNumberGenerator& rng)
+        {
+            const NT eps = epsilon_;
+
+            for (unsigned t = 0; t < walk_len; ++t)
+            {
+                Point v = GetDirection<Point>::apply(P.dimension(), rng);
+
+                NT dot_k = A_row_k_.dot(v.getCoefficients());
+                if (dot_k > NT(0)) {
+                    v     *= NT(-1);
+                    dot_k = -dot_k;
+                }
+
+                const int m_fac = static_cast<int>(P.num_of_hyperplanes());
+                VT Ar(m_fac), Av(m_fac);
```

Comment:
They should both be member variables. Moreover, we shouldn't reset Ar and Av in each iteration. We use them to reduce the cost per step. `P.line_first_positive_intersect(p_, v, Ar, Av, params);` should be called for the first boundary oracle. Then, we should call the boundary oracle that also takes the lambda_hit of the previous step. Plz check Hit-and-Run implementation and mimic the computations there.